torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -149,6 +149,15 @@ ProcessGroupAgent::ProcessGroupAgent(
   }
 }
 
+ProcessGroupAgent::~ProcessGroupAgent() {
+  // TODO: when futures handle failures
+  // (https://github.com/pytorch/pytorch/issues/25531), this should mark
+  // existing futures as completed with an exception.
+  if (listenerThread_.joinable()) {
+    listenerThread_.detach();
```

Comment:
Can we add a log/warning here, as we are living one thread alive? Would I be correct if I assume the listener thread will only die when the process dies?

Reply:
@mrshenli, which thread are we leaving alive? In this case, we detach the listener thread if it is still joinable. This ensures that the thread object is destructed in a valid manner (i.e. it is joined or detached).

Reply:
When does the `listenerThread_` get destructed?

Reply:
The data members of process group agent should be destructed after this destructor body runs. So in this case after this code, `~listenerThread` is called and the `listenerThread` is destructed. It is safe to destroy a thread when the thread was joined or detached. Previously, we would enforce that `join` is called (via `ProcessGroupAgent::join`, so there was no need to possibly detach in the destructor. But now, we check if it has been joined, if not, we detach it, so that the `listenerThread` does not throw an exception (https://en.cppreference.com/w/cpp/thread/thread/~thread)

Reply:
THere is a lint message.

```
Avoid detaching threads and letting them run free; there's a good chance they will access destroyed data after main() exits.  Consider not detaching, and later join()'ing the thread.
```

Does this imply that even if the std::thread data structure is destructed, the POSIX thread`pthread_t ` could still be running?

Reply:
> the POSIX thread pthread_t could still be running?

That's also what I intended to ask. By `listenerThread_`, I actually mean the thread. I am not clear when that thread is destructed, and if we don't feed `recvAnysource` with a message, will that leave it hanging until the process dies? When that happens, will that thread throw any error, and does those errors matter?

Reply:
We should try to shutdown the listenerThread_ as soon as possible and then call join() on it. We can change `listenLoop` to have `while(!shutdown)` and then in the destructor set shutdown (needs to be atomic bool) to true and then call join() on the thread.

We also need to shutdown `threadPool_` cleanly. The reason is that its not safe for the destructor of futureCV_ to run without calling notify on it (https://en.cppreference.com/w/cpp/thread/condition_variable/~condition_variable). Also, its not safe for the destructor of std::mutex to run if a thread is still holding the lock (https://en.cppreference.com/w/cpp/thread/mutex). Note that there is also a case where a thread calling `send` might be holding `futureMutex_` when the destructor runs (maybe we can move that codeblock inside enqueueSend and run it in the threadpool?). In general, we should try to bring down all the threads in ProcessGroupAgent down as quickly as possible and still in a safe manner.

Also, make sure that your destructor doesn't throw an exception: https://isocpp.org/wiki/faq/exceptions#dtors-shouldnt-throw

Reply:
@pritamdamania87 
`listenerThread_` is blocked on calling `pg_->recvAnysource(..)`, there is no good way to make it get back to the loop condition check.

Reply:
> @pritamdamania87
> `listenerThread_` is blocked on calling `pg_->recvAnysource(..)`, there is no good way to make it get back to the loop condition check.

As discussed, this is also why we'd want to be able to have a gloo process send a message to itself.

Reply:
Instead of doing `pg_->recvAnysource(preamble, pg_->getRank())->wait();`, we could do something like this (ideally there should be some way of interrupting pg->recvAnySource):

```
  work = pg_->recvAnysource(preamble, pg_->getRank())
  while (!work.isCompleted() && !shutdown) {
            std::this_thread::sleep_for(
          std::chrono::milliseconds(sleepTime));
  } 
 
 if (shutdown) {
  break;
 }
```

Reply:
> Instead of doing `pg_->recvAnysource(preamble, pg_->getRank())->wait();`, we could do something like this (this is ugly and ideally there should be some way of interrupting pg->recvAnySource):
> 
> ```
>   work = pg_->recvAnysource(preamble, pg_->getRank())
>   while (!work.isCompleted() && !shutdown) {
>             std::this_thread::sleep_for(
>           std::chrono::milliseconds(sleepTime));
>   } 
>  
>  if (shutdown) {
>   break;
>  }
> ```

After some testing, I'm not sure if this approach would work. The `wait` that is called in the current code runs `ProcessGroupGloo::RecvWork::wait`, which sets the `completed_` variable that `isCompleted()` needs to be true to return. Without waiting and blocking on `isCompleted`, there is a forever loop.

Reply:
I was thinking about introducing a ProcessGroupGloo.shutdown method. The shutdown method would do what the destructor does currently. In addition to that it iterates through all contexts in `contexts_` and calls `closeConnections` on each one. Although looking through the gloo code it seems like we hold a lock when we call `read` and `close` needs to hold the same lock, so we might actually end up waiting forever.